no_magic_numbers false positive for enum cases with a negative value

### New Issue Checklist

- [*] Updated SwiftLint to the latest version
- [*] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Negative numbers (except for -1) will trigger the `no_magic_numbers` - e.g. `case negative = -2`. Equivalent positive numbers (e.g. `2`) are fine.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
swiftlint lint Example.swift --enable-all-rules
Linting Swift files at paths Example.swift
Linting 'Example.swift' (1/1)
The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
./Example.swift:3:22: warning: No Magic Numbers Violation: Magic numbers should be replaced by named constants. (no_magic_numbers)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.50.3
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Portable
* Paste your configuration file: None

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using (check `xcodebuild -version`)? 14.2
* Do you have a sample that shows the issue?`echo "internal enum Example {\n    case positive = 2\n    case negative = -2\n}" | swiftlint lint --no-cache --use-stdin --enable-all-rules`


```swift
enum Example: Int {
    case positive = 2
    case negative = -2
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,7 +22,8 @@
   [jimmya](https://github.com/jimmya)
   [#issue_number](https://github.com/realm/SwiftLint/issues/4609)
 
-* Separate analyzer rules as an independent section in the rule directory of the reference.  
+* Separate analyzer rules as an independent section in the rule directory of
+  the reference.  
   [Ethan Wong](https://github.com/GetToSet)
   [#4664](https://github.com/realm/SwiftLint/pull/4664)
 
@@ -47,6 +48,11 @@
   false-positive in if-case-let statements.  
   [SimplyDanny](https://github.com/SimplyDanny)
   [#4548](https://github.com/realm/SwiftLint/issues/4548)
+  
+* Ensure that negative literals in initializers do not trigger
+  `no_magic_numbers` rule.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#4677](https://github.com/realm/SwiftLint/issues/4677)
 
 ## 0.50.3: Bundle of Towels
 

--- a/Source/SwiftLintFramework/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -19,26 +19,32 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
             Example("// array[1337]"),
             Example("baz(\"9999\")"),
             Example("""
-        func foo() {
-            let x: Int = 2
-            let y = 3
-            let vector = [x, y, -1]
-        }
-        """),
-            Example("""
-        class A {
-            var foo: Double = 132
-            static let bar: Double = 0.98
-        }
-        """),
-            Example("""
-        @available(iOS 13, *)
-        func version() {
-            if #available(iOS 13, OSX 10.10, *) {
-                return
+            func foo() {
+                let x: Int = 2
+                let y = 3
+                let vector = [x, y, -1]
             }
-        }
-        """)
+            """),
+            Example("""
+            class A {
+                var foo: Double = 132
+                static let bar: Double = 0.98
+            }
+            """),
+            Example("""
+            @available(iOS 13, *)
+            func version() {
+                if #available(iOS 13, OSX 10.10, *) {
+                    return
+                }
+            }
+            """),
+            Example("""
+            enum Example: Int {
+                case positive = 2
+                case negative = -2
+            }
+            """)
         ],
         triggeringExamples: [
             Example("foo(↓321)"),
@@ -59,13 +65,13 @@ private extension NoMagicNumbersRule {
     final class Visitor: ViolationsSyntaxVisitor {
         override func visitPost(_ node: FloatLiteralExprSyntax) {
             if node.floatingDigits.isMagicNumber {
-                self.violations.append(node.floatingDigits.positionAfterSkippingLeadingTrivia)
+                violations.append(node.floatingDigits.positionAfterSkippingLeadingTrivia)
             }
         }
 
         override func visitPost(_ node: IntegerLiteralExprSyntax) {
             if node.digits.isMagicNumber {
-                self.violations.append(node.digits.positionAfterSkippingLeadingTrivia)
+                violations.append(node.digits.positionAfterSkippingLeadingTrivia)
             }
         }
     }
@@ -73,14 +79,16 @@ private extension NoMagicNumbersRule {
 
 private extension TokenSyntax {
     var isMagicNumber: Bool {
-        let numerStr = text.replacingOccurrences(of: "_", with: "")
-
-        guard let number = Double(numerStr),
-              ![0, 1].contains(number),
-              let parentToken = parent?.parent,
-              !parentToken.is(InitializerClauseSyntax.self) else {
+        guard let number = Double(text.replacingOccurrences(of: "_", with: "")) else {
             return false
         }
-        return true
+        if [0, 1].contains(number) {
+            return false
+        }
+        guard let grandparent = parent?.parent else {
+            return true
+        }
+        return !grandparent.is(InitializerClauseSyntax.self)
+            && grandparent.as(PrefixOperatorExprSyntax.self)?.parent?.is(InitializerClauseSyntax.self) != true
     }
 }

